clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -227,6 +236,103 @@ object GarbageCollector {
                                                spark,
                                                APIConfigurations(apiURL, accessKey, secretKey)
                                               ).withColumn("run_id", lit(runID))
-    expiredAddresses.write.partitionBy("run_id").mode(SaveMode.Append).parquet(addressesDFLocation)
+    expiredAddresses.write
+      .partitionBy("run_id")
+      .mode(SaveMode.Append)
+      .parquet(addressesDFLocation) // TODO(Guys): consider changing to overwrite
+  }
+}
+
+object S3BulkDeleter {
+  def repartitionBySize(df: DataFrame, maxSize: Int, column: String): DataFrame = {
+    val nRows = df.count()
+    val nPartitions = math.max(1, math.floor(nRows / maxSize)).toInt
+    df.repartitionByRange(nPartitions, col(column))
+  }
+
+  def delObjIteration(bucket: String, keys: Seq[String], s3Client: S3Client): Seq[String] = {
+    if (keys.isEmpty) None
+    val removeKeys = keys.map(ObjectIdentifier.builder().key(_).build()).asJava
+    val delObj = Delete.builder().objects(removeKeys).build()
+    val delObjReq = DeleteObjectsRequest.builder.delete(delObj).bucket(bucket).build()
+    val res = s3Client.deleteObjects(delObjReq)
+    res.deleted().asScala.map(_.key())
+  }
+
+  private def getS3Client(region: String, numRetries: Int) = {
+    val retryPolicy = RetryPolicy
+      .builder()
+      .backoffStrategy(BackoffStrategy.defaultThrottlingStrategy())
+      .numRetries(numRetries)
+      .build() // TODO: change to configurable
+    val configuration = ClientOverrideConfiguration.builder().retryPolicy(retryPolicy).build()
+    S3Client.builder.region(Region.of(region)).overrideConfiguration(configuration).build
+  }
+
+  def bulkRemoveFromIter(
+      keys: Iterator[String],
+      bucket: String,
+      region: String,
+      bulkSize: Int,
+      numRetries: Int
+  ): Iterator[mutable.Buffer[String]] = {
+    var nextBatch = keys.take(bulkSize)
+    var res = Seq[mutable.Buffer[String]]()
+    val s3Client = getS3Client(region, numRetries)
+    while (!nextBatch.isEmpty) {
+      res = res ++ delObjIteration(bucket, nextBatch.toSeq, s3Client).map(x =>
+        mutable.Buffer[String](x)
+      )
+      nextBatch = keys.take(bulkSize)
+    }
```

Comment:
A more functional (Scala-ish) way would be to use higher-order operators on a list, e.g. something like
```scala
val s3Client = getS3Client(region, numRetries)
(keys grouped bulkSize).
    flatMap(delObjOperation(bucket, _, s3Client))
```
which it might even be possible to write as
```scala
val s3Client = getS3Client(region, numRetries)
keys grouped bulkSize flatMap delObjOperation(bucket, _, s3Client)
```

Explanation: The operators `grouped` and `flatMap` are defined in https://docs.scala-lang.org/overviews/collections/iterators.html.  Take `keys` in groups of size `bulkSize` (list of batches of keys), run `delObjOperation` on each batch (each batch gives a batch of results), and flatten the list of those batches.